README.md
# ZenWeb Tenant module

[ZenWeb](https://www.npmjs.com/package/zenweb)

多租户多数据库支持

## 安装

```bash
# 生产依赖
npm install @zenweb/tenant @zenweb/mysql @zenweb/repositories zenorm

# 开发依赖
npm install @zenorm/generate @zenorm/generate-mysql --save-dev
```

## 配置

在 `package.json` 的 `scripts` 中增加如下代码，用于执行 `dbgen` 命令

```json title="package.json"
{
  "scripts": {
    "dbgen": "zenorm-generate .dbgen.json"
  }
}
```

创建文件 `.dbgen.json` 用于生成数据库结构代码时连接到指定数据库

*提示：运行时并不使用此配置*

```json title=".dbgen.json"
{
  "host": "localhost",
  "port": 3306,
  "user": "root",
  "password": "",
  "database": "zenorm_test",
  "outputDir": "./model",
  "generateRepositories": true,
  "declareRepositoriesToModules": [
    "@zenweb/core.Core.repositories"
  ]
}
```

### 生成数据库结构代码

运行命令开始生成数据库结构代码
```bash
npm run dbgen
```

## 项目配置

编辑项目入口代码 `src/index.ts`

```ts title="src/index.ts"
import { create } from 'zenweb';
import modTenant, { Tenant } from '@zenweb/tenant';
import { Repositories } from './model';

export const app = create();

/**
 * 租户配置信息
 */
const tenantsConfig: { [id: string]: Tenant } = {
  'a.demo.com': {
    server: 'S1',
    database: 'db_1',
  },
  'b.demo.com': {
    server: 'S2',
    database: 'db_2',
  },
};

app.setup(modTenant({
  tenantGetter: ctx => {
    const tenant = tenantsConfig[ctx.host];
    if (!tenant) {
      throw new Error('Tennat not exists: ' + ctx.host);
    }
    return tenant;
  },
  pools: {
    S1: {
      MASTER: {
        host: '127.0.0.1',
        port: 3306,
        user: 'root',
        password: '123456',
        charset: 'utf8mb4',
        timezone: '+08:00',
        connectionLimit: 100,
      },
    },
    S2: {
      MASTER: {
        host: '127.0.0.1',
        port: 3306,
        user: 'root',
        password: '',
        charset: 'utf8mb4',
        timezone: '+08:00',
        connectionLimit: 100,
      },
    },
  },
  Repositories,
}));
```

代码中调用

```ts title="src/controller/test.ts"
import { mapping, Context } from 'zenweb';

export class TestController {
  @mappping()
  dbtest(ctx: Context) {
    const { UserRepository } = ctx.repositories;
    return UserRepository.find({ id: 1 }).get();
  }
}
```


CHANGELOG.md
# Changelog

## [3.0.0] - 2023-3-24


src/index.ts
import { SetupFunction } from '@zenweb/core';
import modMySQL, { MySQLOption } from '@zenweb/mysql';
import modRepositories from '@zenweb/repositories';
import * as mysql from 'mysql2';
import { Tenant, TenantOption } from './types';
export { Tenant, TenantOption };

/**
 * 当前上下文取得的租户信息
 */
const TENANT = Symbol('tenant');

/**
 * 生成租户数据库配置信息
 */
export function makeTenantMySQLOption(option: TenantOption): MySQLOption {
  // 重组配置
  const pools: { [group: string]: mysql.PoolOptions } = {};
  for (const [server, poolConfig] of Object.entries(option.pools)) {
    pools[`${server}_MASTER`] = poolConfig.MASTER;
    if (poolConfig.SLAVES) {
      let i = 1;
      for (const slave of poolConfig.SLAVES) {
        pools[`${server}_SLAVE${i}`] = slave;
        i++;
      }
    }
  }
  return {
    pools,
    // 使用 Context 初始化连接池
    withContext: true,
    // 所属数据库服务器切换
    getPoolConnectionBeforeWithContext: ctx => {
      return async opt => {
        if (!ctx[TENANT]) {
          ctx[TENANT] = await option.tenantGetter(ctx);
        }
        return {
          // 将原有选择规则增加前缀
          pattern: ctx[TENANT].server + '_' + (opt?.pattern || 'MASTER'),
          selector: opt?.selector,
        }
      };
    },
    // 所属数据库切换 (复用连接池)
    getPoolConnectionAfterWithContext: ctx => {
      return conn => new Promise<void>((resolve, reject) => {
        conn.query('USE `' + ctx[TENANT].database + '`', err => {
          if (err) reject(err);
          else resolve();
        })
      });
    },
  };
}

export default function setup(option: TenantOption): SetupFunction {
  return function tenant(setup) {
    setup.debug('option: %o', option);
    setup.core.setup(modMySQL(makeTenantMySQLOption(option)));
    setup.core.setup(modRepositories({
      contextQuery: ctx => ctx.mysql,
      Repositories: option.Repositories,
      contextProperty: option.contextProperty,
    }));
  }
}

declare module '@zenweb/core' {
  interface Context {
    [TENANT]: Tenant;
  }
}


src/types.ts
import * as mysql from 'mysql2';
import { QueryParam } from 'zenorm';
import { Context } from "@zenweb/core";

/**
 * 租户配置描述
 */
export interface Tenant {
  /**
   * 所属数据库服务器名称
   */
  server: string;

  /**
   * 所属数据库名称
   */
  database: string;
};

/**
 * 租户数据库服务器配置
 */
export interface TenantOption {
  /**
   * 取得租户配置
   * - 如果不能找到请使用 `ctx.fail` 抛出异常，或者返回一个默认租户？
   */
  tenantGetter: (ctx: Context) => Tenant | Promise<Tenant>;

  /**
   * zenorm 生成的 Repositories class
   */
  Repositories: { new (query: QueryParam): void; };

  /**
   * 数据库连接池配置
   */
  pools: {
    /**
     * @param server 所属数据库服务器名称，对应 `Tenant.server`
     */
    [server: string]: {
      /**
       * 主数据库(读写) - 必须设置
       */
      MASTER: mysql.PoolOptions;
      /**
       * 从库(只读) - 可选
       */
      SLAVES?: mysql.PoolOptions[];
    },
  };

  /**
   * 挂载到 `Context` 中的属性名
   * @default 'repositories'
   */
  contextProperty?: string;
}


example/app.ts
import { Core } from '@zenweb/core';
import modTenant, { Tenant } from '../src';
import { Repositories } from './model';

const app = new Core();

/**
 * 租户配置信息
 */
const tenantsConfig: { [id: string]: Tenant } = {
  'a.demo.com': {
    server: 'S1',
    database: 'db_1',
  },
  'b.demo.com': {
    server: 'S2',
    database: 'db_2',
  },
};

app.setup(modTenant({
  tenantGetter: ctx => {
    const tenant = tenantsConfig[ctx.host];
    if (!tenant) {
      throw new Error('Tennat not exists: ' + ctx.host);
    }
    return tenant;
  },
  pools: {
    S1: {
      MASTER: {
        host: '127.0.0.1',
        port: 3306,
        user: 'root',
        password: '123456',
        charset: 'utf8mb4',
        timezone: '+08:00',
        connectionLimit: 100,
      },
    },
    S2: {
      MASTER: {
        host: '127.0.0.1',
        port: 3306,
        user: 'root',
        password: '',
        charset: 'utf8mb4',
        timezone: '+08:00',
        connectionLimit: 100,
      },
    },
  },
  Repositories,
}));

app.setup(function test(setup) {
  setup.middleware(async ctx => {
    const { UserRepository } = ctx.repositories;
    ctx.body = await UserRepository.of('MASTER').find().all();
  })
});

app.start();
